File: agent-app-go/internal/core/evidence_optimizer.go
package core

import (
	"regexp"
	"sort"
	"strconv"
	"strings"

	agentcontext "github.com/riwonace/agent-app-go/internal/context"
	"github.com/riwonace/agent-app-go/internal/router"
)

const (
	MinRelevance      = 0.3
	DuplicateThreshold = 0.8
)

var reliabilityWeights = map[string]float64{
	"sql":             1.0,
	"knowledge-graph": 0.9,
	"document":        0.7,
	"vector":          0.6,
}

var hasDigitRe = regexp.MustCompile(`\d+`)

// OptimizationResult mirrors data class OptimizationResult.
type OptimizationResult struct {
	Selected          []agentcontext.ContextItem
	TotalValue        float64
	UsedBudget        int
	ExcludedCount     int
	DeduplicatedCount int
	Reasoning         string
}

// EvidenceOptimizer ports EvidenceOptimizer.kt 1:1: greedy value/cost evidence selection
// within a character budget.
type EvidenceOptimizer struct {
	BudgetChars int
}

type scoredItem struct {
	item agentcontext.ContextItem
	value float64
	cost  int
}

func (o *EvidenceOptimizer) Optimize(items []agentcontext.ContextItem, routes []router.Route, profile QueryProfile) OptimizationResult {
	if len(items) == 0 {
		return OptimizationResult{Reasoning: "증거 항목 없음"}
	}

	var filtered []agentcontext.ContextItem
	for _, it := range items {
		if it.Score >= MinRelevance {
			filtered = append(filtered, it)
		}
	}
	filteredOutCount := len(items) - len(filtered)

	deduplicated := deduplicateItems(filtered)
	deduplicatedCount := len(filtered) - len(deduplicated)

	scored := make([]scoredItem, 0, len(deduplicated))
	for _, it := range deduplicated {
		scored = append(scored, scoredItem{
			item:  it,
			value: calculateValue(it, routes, profile),
			cost:  len([]rune(it.Text)),
		})
	}

	sort.SliceStable(scored, func(i, j int) bool {
		ci := scored[i].cost
		if ci < 1 {
			ci = 1
		}
		cj := scored[j].cost
		if cj < 1 {
			cj = 1
		}
		return scored[i].value/float64(ci) > scored[j].value/float64(cj)
	})

	var selected []scoredItem
	remainingBudget := effectiveBudget(routes, o.BudgetChars)

	for _, s := range scored {
		if s.cost <= remainingBudget {
			selected = append(selected, s)
			remainingBudget -= s.cost
		} else if len(selected) == 0 {
			runes := []rune(s.item.Text)
			cut := remainingBudget - 12
			if cut < 0 {
				cut = 0
			}
			if cut > len(runes) {
				cut = len(runes)
			}
			truncatedText := string(runes[:cut]) + "…(truncated)"
			truncated := scoredItem{
				item: agentcontext.ContextItem{Source: s.item.Source, Text: truncatedText, Score: s.item.Score},
				value: s.value,
				cost:  remainingBudget,
			}
			selected = append(selected, truncated)
			remainingBudget = 0
			break
		}
	}

	selectedItems := make([]agentcontext.ContextItem, 0, len(selected))
	for _, s := range selected {
		selectedItems = append(selectedItems, s.item)
	}
	finalSelected := applyLostInMiddleMitigation(selectedItems)

	totalValue := 0.0
	usedBudget := 0
	for _, s := range selected {
		totalValue += s.value
		usedBudget += s.cost
	}

	return OptimizationResult{
		Selected:          finalSelected,
		TotalValue:        totalValue,
		UsedBudget:        usedBudget,
		ExcludedCount:     filteredOutCount + (len(deduplicated) - len(selected)),
		DeduplicatedCount: deduplicatedCount,
		Reasoning:         buildOptimizerReasoning(len(selected), filteredOutCount, deduplicatedCount, totalValue),
	}
}

func calculateValue(item agentcontext.ContextItem, routes []router.Route, profile QueryProfile) float64 {
	relevance := item.Score
	reliability, ok := reliabilityWeights[item.Source]
	if !ok {
		reliability = 0.5
	}
	answerability := calculateAnswerability(item, profile)
	novelty := calculateNovelty(item, routes)
	return relevance * reliability * answerability * novelty
}

func calculateAnswerability(item agentcontext.ContextItem, profile QueryProfile) float64 {
	score := 1.0
	switch profile.Intent {
	case IntentFactual, IntentAggregation:
		if hasDigitRe.MatchString(item.Text) {
			score *= 1.2
		}
		if item.Source == "sql" {
			score *= 1.3
		}
	case IntentComparison:
		if countDigits(item.Text) >= 2 {
			score *= 1.2
		}
	case IntentExplanation:
		if len([]rune(item.Text)) > 200 {
			score *= 1.1
		}
	case IntentRelation:
		if item.Source == "knowledge-graph" {
			score *= 1.3
		}
	case IntentMultiHop:
		score *= 1.0
	}
	if score > 1.5 {
		return 1.5
	}
	return score
}

func countDigits(s string) int {
	n := 0
	for _, r := range s {
		if r >= '0' && r <= '9' {
			n++
		}
	}
	return n
}

func calculateNovelty(item agentcontext.ContextItem, routes []router.Route) float64 {
	sourceToRoute := map[string]router.Route{
		"sql":             router.RouteSQL,
		"knowledge-graph": router.RouteGraph,
		"document":        router.RouteVector,
		"vector":          router.RouteVector,
	}
	itemRoute, ok := sourceToRoute[item.Source]
	if ok {
		for _, r := range routes {
			if r == itemRoute {
				return 1.2
			}
		}
	}
	return 0.8
}

var wsRe = regexp.MustCompile(`\s+`)

func deduplicateItems(items []agentcontext.ContextItem) []agentcontext.ContextItem {
	if len(items) <= 1 {
		return items
	}
	sorted := append([]agentcontext.ContextItem{}, items...)
	sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].Score > sorted[j].Score })

	seen := make(map[string]bool)
	var result []agentcontext.ContextItem
	for _, it := range sorted {
		normalized := strings.TrimSpace(wsRe.ReplaceAllString(it.Text, " "))
		runes := []rune(normalized)
		if len(runes) > 100 {
			runes = runes[:100]
		}
		hash := string(runes)
		if !seen[hash] {
			seen[hash] = true
			result = append(result, it)
		}
	}
	return result
}

func applyLostInMiddleMitigation(items []agentcontext.ContextItem) []agentcontext.ContextItem {
	if len(items) < 3 {
		return items
	}
	result := append([]agentcontext.ContextItem{}, items...)
	second := result[1]
	result = append(result[:1], result[2:]...)
	result = append(result, second)
	return result
}

func effectiveBudget(routes []router.Route, budgetChars int) int {
	if len(routes) == 1 && routes[0] == router.RouteSQL {
		return budgetChars / 2
	}
	return budgetChars
}

func buildOptimizerReasoning(selectedCount, filteredOut, deduplicated int, totalValue float64) string {
	parts := []string{strconv.Itoa(selectedCount) + "개 증거 선택"}
	if filteredOut > 0 {
		parts = append(parts, "관련도 미달 "+strconv.Itoa(filteredOut)+"개 제외")
	}
	if deduplicated > 0 {
		parts = append(parts, "중복 "+strconv.Itoa(deduplicated)+"개 제거")
	}
	return strings.Join(parts, ", ")
}

File: agent-app-ktor/src/main/kotlin/com/riwonace/agentktor/core/EvidenceOptimizer.kt
package com.riwonace.agentktor.core

import com.riwonace.agentktor.context.ContextItem
import com.riwonace.agentktor.router.Route
import org.slf4j.LoggerFactory
import kotlin.math.min

/**
 * Evidence Budget Optimizer: 토큰 예산 내에서 최적의 증거 조합을 선택한다.
 *
 * 기존 ContextCurator의 단순 예산 기반 선택에서
 * 제약 최적화 문제로 전환:
 *
 * maximize: Σ(relevance × reliability × answerability × novelty)
 * subject to: Σ(token_cost) ≤ budget
 *
 * 그리디 알고리즘으로 근사해를 구한다.
 */
class EvidenceOptimizer(
    private val budgetChars: Int = 2400,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    companion object {
        /** 최소 관련도 임계값 */
        const val MIN_RELEVANCE = 0.3

        /** 중복 판정 유사도 임계값 */
        const val DUPLICATE_THRESHOLD = 0.8

        /** 소스별 신뢰도 가중치 */
        val RELIABILITY_WEIGHTS = mapOf(
            "sql" to 1.0,           // SQL 결과는 신뢰도 최고
            "knowledge-graph" to 0.9,
            "document" to 0.7,
            "vector" to 0.6,
        )
    }

    /**
     * 최적화된 증거 선택 결과
     */
    data class OptimizationResult(
        /** 선택된 증거 항목 */
        val selected: List<ContextItem>,
        /** 선택된 항목의 총 가치 */
        val totalValue: Double,
        /** 사용된 예산 (문자 수) */
        val usedBudget: Int,
        /** 제외된 항목 수 */
        val excludedCount: Int,
        /** 중복으로 제거된 항목 수 */
        val deduplicatedCount: Int,
        /** 최적화 사유 */
        val reasoning: String,
    )

    /**
     * 증거 항목들을 최적화하여 예산 내 최적 조합을 반환한다.
     *
     * @param items 후보 증거 항목들
     * @param routes 선택된 라우트들 (라우트별 가중치 적용)
     * @param profile 질문 프로파일 (answerability 계산용)
     */
    fun optimize(
        items: List<ContextItem>,
        routes: List<Route>,
        profile: QueryProfile,
    ): OptimizationResult {
        if (items.isEmpty()) {
            return OptimizationResult(
                selected = emptyList(),
                totalValue = 0.0,
                usedBudget = 0,
                excludedCount = 0,
                deduplicatedCount = 0,
                reasoning = "증거 항목 없음",
            )
        }

        // 1. 최소 관련도 필터링
        val filtered = items.filter { it.score >= MIN_RELEVANCE }
        val filteredOutCount = items.size - filtered.size

        // 2. 중복 제거
        val deduplicated = deduplicateItems(filtered)
        val deduplicatedCount = filtered.size - deduplicated.size

        // 3. 가치 점수 계산
        val scoredItems = deduplicated.map { item ->
            ScoredItem(
                item = item,
                value = calculateValue(item, routes, profile),
                cost = item.text.length,
            )
        }

        // 4. 효율성 기준 정렬 (가치/비용 비율)
        val sortedByEfficiency = scoredItems.sortedByDescending { it.value / it.cost.coerceAtLeast(1) }

        // 5. 그리디 선택
        val selected = mutableListOf<ScoredItem>()
        var remainingBudget = calculateEffectiveBudget(routes)

        for (scored in sortedByEfficiency) {
            if (scored.cost <= remainingBudget) {
                selected.add(scored)
                remainingBudget -= scored.cost
            } else if (selected.isEmpty()) {
                // 첫 항목이 예산 초과면 잘라서라도 포함
                val truncated = scored.copy(
                    item = scored.item.copy(
                        text = scored.item.text.take(remainingBudget - 12) + "…(truncated)",
                    ),
                    cost = remainingBudget,
                )
                selected.add(truncated)
                remainingBudget = 0
                break
            }
        }

        // 6. Lost in the Middle 완화: 3개 이상이면 차상위를 마지막으로
        val finalSelected = applyLostInMiddleMitigation(selected.map { it.item })

        val totalValue = selected.sumOf { it.value }
        val usedBudget = selected.sumOf { it.cost }

        log.info(
            "Evidence 최적화: {} 항목 → {} 선택 (가치={}, 예산={})",
            items.size, finalSelected.size, "%.2f".format(totalValue), usedBudget,
        )

        return OptimizationResult(
            selected = finalSelected,
            totalValue = totalValue,
            usedBudget = usedBudget,
            excludedCount = filteredOutCount + (deduplicated.size - selected.size),
            deduplicatedCount = deduplicatedCount,
            reasoning = buildReasoning(selected.size, filteredOutCount, deduplicatedCount, totalValue),
        )
    }

    /**
     * 증거 항목의 가치 점수 계산
     *
     * value = relevance × reliability × answerability × novelty
     */
    private fun calculateValue(
        item: ContextItem,
        routes: List<Route>,
        profile: QueryProfile,
    ): Double {
        // 1. Relevance: 기존 score 활용
        val relevance = item.score

        // 2. Reliability: 소스별 신뢰도
        val reliability = RELIABILITY_WEIGHTS[item.source] ?: 0.5

        // 3. Answerability: 질문 의도에 부합하는 정도
        val answerability = calculateAnswerability(item, profile)

        // 4. Novelty: 라우트 가중치 (선택된 라우트와 매칭되면 보너스)
        val novelty = calculateNovelty(item, routes)

        return relevance * reliability * answerability * novelty
    }

    /**
     * 답변 가능성 점수 계산
     */
    private fun calculateAnswerability(item: ContextItem, profile: QueryProfile): Double {
        var score = 1.0

        when (profile.intent) {
            QueryIntent.FACTUAL, QueryIntent.AGGREGATION -> {
                // 숫자가 포함된 증거 선호
                if (item.text.contains(Regex("\\d+"))) score *= 1.2
                // SQL 결과 선호
                if (item.source == "sql") score *= 1.3
            }
            QueryIntent.COMPARISON -> {
                // 비교 가능한 데이터 선호
                if (item.text.count { it.isDigit() } >= 2) score *= 1.2
            }
            QueryIntent.EXPLANATION -> {
                // 긴 설명 텍스트 선호
                if (item.text.length > 200) score *= 1.1
            }
            QueryIntent.RELATION -> {
                // 관계 키워드 포함 선호
                if (item.source == "knowledge-graph") score *= 1.3
            }
            QueryIntent.MULTI_HOP -> {
                // 다양한 정보 선호
                score *= 1.0
            }
        }

        return score.coerceAtMost(1.5)
    }

    /**
     * 신규성 점수 계산 (라우트 매칭)
     */
    private fun calculateNovelty(item: ContextItem, routes: List<Route>): Double {
        val sourceToRoute = mapOf(
            "sql" to Route.SQL,
            "knowledge-graph" to Route.GRAPH,
            "document" to Route.VECTOR,
            "vector" to Route.VECTOR,
        )

        val itemRoute = sourceToRoute[item.source]
        return if (itemRoute != null && itemRoute in routes) 1.2 else 0.8
    }

    /**
     * 중복 항목 제거
     */
    private fun deduplicateItems(items: List<ContextItem>): List<ContextItem> {
        if (items.size <= 1) return items

        val result = mutableListOf<ContextItem>()
        val seen = mutableSetOf<String>()

        for (item in items.sortedByDescending { it.score }) {
            val normalized = item.text.replace(Regex("\\s+"), " ").trim()
            val hash = normalized.take(100) // 앞 100자로 중복 판정

            if (hash !in seen) {
                seen.add(hash)
                result.add(item)
            }
        }

        return result
    }

    /**
     * Lost in the Middle 완화: 3개 이상이면 차상위를 마지막으로
     */
    private fun applyLostInMiddleMitigation(items: List<ContextItem>): List<ContextItem> {
        if (items.size < 3) return items

        val result = items.toMutableList()
        val second = result.removeAt(1)
        result.add(second)
        return result
    }

    /**
     * 유효 예산 계산 (SQL 단독이면 절반)
     */
    private fun calculateEffectiveBudget(routes: List<Route>): Int {
        return if (routes == listOf(Route.SQL)) {
            budgetChars / 2
        } else {
            budgetChars
        }
    }

    private fun buildReasoning(
        selectedCount: Int,
        filteredOut: Int,
        deduplicated: Int,
        totalValue: Double,
    ): String {
        val parts = mutableListOf<String>()
        parts += "${selectedCount}개 증거 선택"
        if (filteredOut > 0) parts += "관련도 미달 ${filteredOut}개 제외"
        if (deduplicated > 0) parts += "중복 ${deduplicated}개 제거"
        parts += "총 가치 ${String.format("%.2f", totalValue)}"
        return parts.joinToString(", ")
    }

    private data class ScoredItem(
        val item: ContextItem,
        val value: Double,
        val cost: Int,
    )
}
